File: src/ios/ImageFilter.m
//
//  ImageFilter.m
//  HelloWorld
//
//  Created by dat tran on 3/3/17.
//
//

#import "ImageFilter.h"

@implementation ImageFilter

static CIContext *context;
static UIImage *currentEditingImage;
static UIImage *currentPreviewImage;
static UIImage *currentThumbnailImage;
static NSString *currentImagePath = nil;
static NSString *base64Image = nil;
static CGSize screenSize;


static NSString* toBase64(NSData* data) {
    SEL s1 = NSSelectorFromString(@"cdv_base64EncodedString");
    SEL s2 = NSSelectorFromString(@"base64EncodedString");
    SEL s3 = NSSelectorFromString(@"base64EncodedStringWithOptions:");

    if ([data respondsToSelector:s1]) {
        NSString* (*func)(id, SEL) = (void *)[data methodForSelector:s1];
        return func(data, s1);
    } else if ([data respondsToSelector:s2]) {
        NSString* (*func)(id, SEL) = (void *)[data methodForSelector:s2];
        return func(data, s2);
    } else if ([data respondsToSelector:s3]) {
        NSString* (*func)(id, SEL, NSUInteger) = (void *)[data methodForSelector:s3];
        return func(data, s3, 0);
    } else {
        return nil;
    }
}

static UIImage * base64ToImage(NSString *base64Image) {
    NSData *imageData = [[NSData alloc] initWithBase64EncodedString:base64Image options:0];
    return [UIImage imageWithData:imageData];
}

- (void)pluginInitialize {
    [super pluginInitialize];

    EAGLContext *eaglCxt = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
    context = [CIContext contextWithEAGLContext:eaglCxt];

    screenSize = [[UIScreen mainScreen] bounds].size;
}

- (void)validateInput:(CDVInvokedUrlCommand *)command {
    NSString *pathOrData = [command argumentAtIndex:0 withDefault:nil];
    NSURL *pathUrl = [NSURL URLWithString:pathOrData];
    pathOrData = pathUrl.path;
    NSString *filterType = [command argumentAtIndex:1 withDefault:nil];
    NSNumber *isBase64Image = [command argumentAtIndex:3 withDefault:@(0)];

    if ([isBase64Image intValue] == 0)
    {
        if (pathOrData.length && filterType.length && ![currentImagePath isEqualToString:pathOrData]) {
            currentImagePath = pathOrData;
            currentEditingImage = [UIImage imageWithContentsOfFile:pathOrData];

            float ratio = screenSize.width / currentEditingImage.size.width;
            CGSize newSize = CGSizeMake(currentEditingImage.size.width * ratio, currentEditingImage.size.height * ratio);
            UIGraphicsBeginImageContextWithOptions(newSize, NO, 0.0);
            [currentEditingImage drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
            currentPreviewImage = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();

            CGSize thumbnailSize = CGSizeMake(currentPreviewImage.size.width * 0.2f, currentPreviewImage.size.height * 0.2f);
            UIGraphicsBeginImageContextWithOptions(thumbnailSize, NO, 0.0);
            [currentPreviewImage drawInRect:CGRectMake(0, 0, thumbnailSize.width, thumbnailSize.height)];
            currentThumbnailImage = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
        }
    }
    else if ([isBase64Image intValue] == 1) {
        if (pathOrData.length && filterType.length && ![pathOrData isEqualToString:base64Image]) {
            base64Image = pathOrData;
            currentEditingImage = base64ToImage(pathOrData);

            float ratio = screenSize.width / currentEditingImage.size.width;
            CGSize newSize = CGSizeMake(currentEditingImage.size.width * ratio, currentEditingImage.size.height * ratio);
            UIGraphicsBeginImageContextWithOptions(newSize, NO, 0.0);
            [currentEditingImage drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
            currentPreviewImage = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();

            CGSize thumbnailSize = CGSizeMake(currentPreviewImage.size.width * 0.2f, currentPreviewImage.size.height * 0.2f);
            UIGraphicsBeginImageContextWithOptions(thumbnailSize, NO, 0.0);
            [currentPreviewImage drawInRect:CGRectMake(0, 0, thumbnailSize.width, thumbnailSize.height)];
            currentThumbnailImage = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
        }
    }
    else {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

- (void)applyEffect:(CDVInvokedUrlCommand*)command {
    [self validateInput:command];

    NSString *filterType = [command argumentAtIndex:1 withDefault:nil];
    NSNumber *compressionQuality = [command argumentAtIndex:2 withDefault:@(0.5)];

    [self filterImage:currentEditingImage filter:filterType compressionQuality:compressionQuality completion:^(NSData *data) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:toBase64(data)];
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        });
    }];
}

- (void)applyEffectForReview:(CDVInvokedUrlCommand*)command {
    [self validateInput:command];

    NSString *filterType = [command argumentAtIndex:1 withDefault:nil];
    NSNumber *compressionQuality = [command argumentAtIndex:2 withDefault:@(0.5)];

    [self filterImage:currentPreviewImage filter:filterType compressionQuality:compressionQuality completion:^(NSData *data) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:toBase64(data)];
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        });
    }];
}

- (void)applyEffectForThumbnail:(CDVInvokedUrlCommand *)command {
    [self validateInput:command];

    NSString *filterType = [command argumentAtIndex:1 withDefault:nil];
    NSNumber *compressionQuality = [command argumentAtIndex:2 withDefault:@(0.5)];

    [self filterImage:currentThumbnailImage filter:filterType compressionQuality:compressionQuality completion:^(NSData *data) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:toBase64(data)];
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        });
    }];
}

- (void)filterImage:(UIImage *)image filter:(NSString *)filterType compressionQuality:(NSNumber *)quality completion:(void(^)(NSData *))completion {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        @autoreleasepool {
            UIImage *result = [self applySelectedEffect:image effect:filterType];
            NSData *data = UIImageJPEGRepresentation(result, [quality floatValue]);
            completion(data);
        }
    });
}

- (UIImage *)applySelectedEffect:(UIImage *)image effect:(NSString *)effect {
    if ([effect isEqualToString:@"aged"])
        return [self applyAgedEffect:image];
    else if ([effect isEqualToString:@"blackWhite"])
        return [self applyBlackWhiteEffect:image];
    else if ([effect isEqualToString:@"cold"])
        return [self applyColdEffect:image];
    else if ([effect isEqualToString:@"rosy"])
        return [self applyRosyEffect:image];
    else if ([effect isEqualToString:@"intense"])
        return [self applyIntenseEffect:image];
    else if ([effect isEqualToString:@"warm"])
        return [self applyWarmEffect:image];
    else if ([effect isEqualToString:@"light"])
        return [self applyLightEffect:image];
    else
        return nil;
}

- (UIImage *)applyAgedEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(0.00516)
                  brightness:@(0.04124)
                    contrast:@(0.8763)
                       gamma:@(0.7474)
                    exposure:@(0.1804)
                     sharpen:@(1.0103)
                         hue:nil
                         red:@(0.7835)
                       green:@(0.719)
                        blue:@(0.616)];
}

- (UIImage *)applyBlackWhiteEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(0.0f)
                  brightness:nil
                    contrast:nil
                       gamma:@(1.2282)
                    exposure:@(0.2062)
                     sharpen:@(0.268)
                         hue:nil
                         red:nil
                       green:nil
                        blue:nil];
}

- (UIImage *)applyColdEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(0.216495)
                  brightness:@(-0.134021)
                    contrast:@(0.85567)
                       gamma:@(1.061856)
                    exposure:@(0.603093)
                     sharpen:nil
                         hue:nil
                         red:@(0.708763)
                       green:@(0.832474)
                        blue:nil];
}

- (UIImage *)applyRosyEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(0.79897)
                  brightness:@(-0.164948)
                    contrast:@(0.819588)
                       gamma:@(0.881443)
                    exposure:@(0.474227)
                     sharpen:nil
                         hue:nil
                         red:nil
                       green:@(0.822165)
                        blue:@(0.876289)];
}

- (UIImage *)applyIntenseEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(2.0)
                  brightness:nil
                    contrast:nil
                       gamma:nil
                    exposure:@(0.12371)
                     sharpen:nil
                         hue:nil
                         red:nil
                       green:nil
                        blue:nil];
}

- (UIImage *)applyWarmEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(1.2577)
                  brightness:@(-0.085)
                    contrast:@(0.964)
                       gamma:@(0.8763)
                    exposure:@(0.4536)
                     sharpen:nil
                         hue:nil
                         red:@(0.83)
                       green:@(0.8092)
                        blue:@(0.7938)];
}

- (UIImage *)applyLightEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:@(1.4484)
                  brightness:@(-0.0592)
                    contrast:@(0.7629)
                       gamma:@(0.7835)
                    exposure:@(0.4124)
                     sharpen:@(-0.0825)
                         hue:nil
                         red:nil
                       green:nil
                        blue:nil];
}

- (UIImage *)applyNoEffect:(UIImage *)img {
    return [self applyFilter:img
                  saturation:nil
                  brightness:nil
                    contrast:nil
                       gamma:nil
                    exposure:nil
                     sharpen:nil
                         hue:nil
                         red:nil
                       green:nil
                        blue:nil];
}

- (UIImage *)applyFilter:(UIImage *)image saturation:(NSNumber *)saturation brightness:(NSNumber *)brightness contrast:(NSNumber *)contrast gamma:(NSNumber *)gamma exposure:(NSNumber *)exposure sharpen:(NSNumber *)sharpen hue:(NSNumber *)hue red:(NSNumber *)red green:(NSNumber *)green blue:(NSNumber *)blue {

    UIImage *result;
    NSMutableArray<GPUImageFilter*> *filters = [NSMutableArray new];
    GPUImageFilter *tmp;
    if (saturation) {
        GPUImageFilter *filter = [self applySaturation:result saturation:saturation];
        [filters addObject:filter];
        tmp = filter;
    }
    if (brightness) {
        GPUImageFilter *filter = [self applyBrightness:result brightness:brightness];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (contrast) {
        GPUImageFilter *filter = [self applyContrast:result contrast:contrast];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (gamma){
        GPUImageFilter *filter = [self applyGamma:result value:gamma];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (exposure){
        GPUImageFilter *filter = [self applyExposure:result value:exposure];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (sharpen){
        GPUImageFilter *filter = [self applySharpen:result value:sharpen];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (hue){
        GPUImageFilter *filter = [self applyHue:result value:hue];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (red){
        GPUImageFilter *filter = [self applyRed:result value:red];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (green){
        GPUImageFilter *filter = [self applyGreen:result value:green];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }
    if (blue){
        GPUImageFilter *filter = [self applyBlue:result value:blue];
        [filters addObject:filter];
        [tmp addTarget:filter];
        tmp = filter;
    }

    GPUImagePicture *stillImageSource = [[GPUImagePicture alloc] initWithImage:image];
    GPUImageFilterGroup *group = [GPUImageFilterGroup new];
    [group setInitialFilters:[NSArray arrayWithObject:filters.firstObject]];
    [group setTerminalFilter:filters.lastObject];

    [stillImageSource addTarget:group];
    [group useNextFrameForImageCapture];
    [stillImageSource processImage];

    result = [group imageFromCurrentFramebuffer];

    return result;
}

- (GPUImageFilter *)applyBrightness:(UIImage *)image brightness:(NSNumber*)brightness {
    GPUImageBrightnessFilter *filter = [GPUImageBrightnessFilter new];
    [filter setBrightness:[brightness floatValue]];
    return filter;

}
- (GPUImageFilter *)applySaturation:(UIImage *)image saturation:(NSNumber*)saturation {
    GPUImageSaturationFilter *filter = [GPUImageSaturationFilter new];
    [filter setSaturation:[saturation floatValue]];
    return filter;
}
- (GPUImageFilter *)applyContrast:(UIImage *)image contrast:(NSNumber*)contrast {
    GPUImageContrastFilter *filter = [GPUImageContrastFilter new];
    [filter setContrast:[contrast floatValue]];
    return filter;
}

- (GPUImageFilter *)applyExposure:(UIImage *)image value:(NSNumber*)value {
    GPUImageExposureFilter *filter = [GPUImageExposureFilter new];
    [filter setExposure:[value floatValue]];
    return filter;
}

- (GPUImageFilter *)applyGamma:(UIImage *)image value:(NSNumber*)value {
    GPUImageGammaFilter *filter = [GPUImageGammaFilter new];
    [filter setGamma:[value floatValue]];
    return filter;
}

- (GPUImageFilter *)applySharpen:(UIImage *)image value:(NSNumber*)value {
    GPUImageSharpenFilter *filter = [GPUImageSharpenFilter new];
    [filter setSharpness:[value floatValue]];
    return filter;
}

- (GPUImageFilter *)applyHue:(UIImage *)image value:(NSNumber*)value {
    GPUImageHueFilter *filter = [GPUImageHueFilter new];
    [filter setHue:[value floatValue]];
    return filter;
}

- (GPUImageFilter*)applyRed:(UIImage*)image value:(NSNumber *)value {
    GPUImageRGBFilter *filter = [GPUImageRGBFilter new];
    [filter setRed:[value floatValue]];
    return filter;
}
- (GPUImageFilter*)applyGreen:(UIImage*)image value:(NSNumber *)value {
    GPUImageRGBFilter *filter = [GPUImageRGBFilter new];
    [filter setGreen:[value floatValue]];
    return filter;
}
- (GPUImageFilter*)applyBlue:(UIImage*)image value:(NSNumber *)value {
    GPUImageRGBFilter *filter = [GPUImageRGBFilter new];
    [filter setBlue:[value floatValue]];
    return filter;
}

@end

File: src/android/ImageFilter.java
package cordova.plugin.sts.image.filter;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Base64;

import org.apache.commons.lang3.StringUtils;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;

import android.graphics.Bitmap.*;
import android.util.DisplayMetrics;

import jp.co.cyberagent.android.gpuimage.*;

/**
 * Created by Dat Tran on 3/7/17.
 */

//private class GPUImageGLSurfaceView extends GLSurfaceView {
//    public GPUImageGLSurfaceView(Context context) {
//        super(context);
//    }
//
//    public GPUImageGLSurfaceView(Context context, AttributeSet attrs) {
//        super(context, attrs);
//    }
//
//    @Override
//    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
//        if (mForceSize != null) {
//            super.onMeasure(MeasureSpec.makeMeasureSpec(mForceSize.width, MeasureSpec.EXACTLY),
//                    MeasureSpec.makeMeasureSpec(mForceSize.height, MeasureSpec.EXACTLY));
//        } else {
//            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
//        }
//    }
//}

class MySize {
    public MySize(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public int getHeight() {
        return height;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public int getWidth() {
        return width;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    private int width;
    private int height;
}

public class ImageFilter extends CordovaPlugin {

    private static final int JPEG = 0;
    private static final int PNG = 1;

    private static final float NOT_AVAILABLE = -9999;

    private static String currentImagePath;
    private static String base64Image;
    private Bitmap currentPreviewImage;
    private Bitmap currentEditingImage;
    private Bitmap currentThumbnailImage;
//    private static GPUImage editingGPUImage;
//    private static GPUImage previewGPUImage;
//    private static GPUImage thumbnailGPUImage;

    //    private GLSurfaceView glSurfaceView;
    private static MySize screenSize;

    private Context context;

    public CallbackContext callbackContext;

    @Override
    protected void pluginInitialize() {
        super.pluginInitialize();

        context = this.cordova.getActivity();

        DisplayMetrics displayMetrics = new DisplayMetrics();
        this.cordova.getActivity().getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        int height = displayMetrics.heightPixels;
        int width = displayMetrics.widthPixels;
        screenSize = new MySize(width, height);

//        glSurfaceView = new GLSurfaceView(context);

//        editingGPUImage.setGLSurfaceView(glSurfaceView);
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {

        this.callbackContext = callbackContext;

        if (action.equals("coolMethod")) {
            String message = args.getString(0);
//            this.coolMethod(message, callbackContext);
            return true;
        }
        if (action.equals("applyEffect")
                || action.equals("applyEffectForReview")
                || action.equals("applyEffectForThumbnail")) {
            String path = args.getString(0);
            String filterType = args.getString(1);
            double compressQuality = args.getDouble(2);
            int isBase64Image = args.getInt(3);

            if (action.equals("applyEffect"))
                this.applyEffect(path, filterType, compressQuality, isBase64Image, callbackContext);
            else if (action.equals("applyEffectForReview"))
                this.applyEffectForReview(path, filterType, compressQuality, isBase64Image, callbackContext);
            else if (action.equals("applyEffectForThumbnail"))
                this.applyEffectForThumbnail(path, filterType, compressQuality, isBase64Image, callbackContext);

            return true;
        }
        return false;
    }

    private void validateInput(String pathOrData, String filterType, int isBase64Image) {

        if (isBase64Image == 0) {
            if (!StringUtils.isEmpty(pathOrData) && !StringUtils.isEmpty(filterType)
                    && !pathOrData.equals(currentImagePath)) {

                currentImagePath = pathOrData;
                pathOrData = pathOrData.substring(7);

                currentEditingImage = BitmapFactory.decodeFile(pathOrData);

                float ratio = (float) screenSize.getWidth() / (float) currentEditingImage.getWidth();
                MySize newSize = new MySize(Math.round(currentEditingImage.getWidth() * ratio), Math.round(currentEditingImage.getHeight() * ratio));
                currentPreviewImage = Bitmap.createScaledBitmap(currentEditingImage, newSize.getWidth(), newSize.getHeight(), false);

                MySize thumbSize = new MySize(Math.round(currentPreviewImage.getWidth() * 0.2f), Math.round(currentPreviewImage.getHeight() * 0.2f));
                currentThumbnailImage = Bitmap.createScaledBitmap(currentPreviewImage, thumbSize.getWidth(), thumbSize.getHeight(), false);
            }
        } else if (isBase64Image == 1) {
            if (!StringUtils.isEmpty(pathOrData) && !StringUtils.isEmpty(filterType)
                    && !pathOrData.equals(base64Image)) {
                base64Image = pathOrData;

                currentEditingImage = base64ToBitmap(pathOrData);

                float ratio = (float) screenSize.getWidth() / (float) currentEditingImage.getWidth();
                MySize newSize = new MySize(Math.round(currentEditingImage.getWidth() * ratio), Math.round(currentEditingImage.getHeight() * ratio));
                currentPreviewImage = Bitmap.createScaledBitmap(currentEditingImage, newSize.getWidth(), newSize.getHeight(), false);

                MySize thumbSize = new MySize(Math.round(currentPreviewImage.getWidth() * 0.2f), Math.round(currentPreviewImage.getHeight() * 0.2f));
                currentThumbnailImage = Bitmap.createScaledBitmap(currentPreviewImage, thumbSize.getWidth(), thumbSize.getHeight(), false);
            }
        } else {
            this.callbackContext.error("something wrong while passing isBase64Image value");
        }
    }

    private Bitmap getThumbnailImage() {
        MySize thumbSize = new MySize(Math.round(currentPreviewImage.getWidth() * 0.2f), Math.round(currentPreviewImage.getHeight() * 0.2f));
        return Bitmap.createScaledBitmap(currentPreviewImage, thumbSize.getWidth(), thumbSize.getHeight(), false);
    }

    private void applyEffect(String pathOrData, final String filterType, final double compressQuality, int isBase64Image, CallbackContext callbackContext) {
        synchronized (this) {
            this.validateInput(pathOrData, filterType, isBase64Image);

            Bitmap bmp = null;
            GPUImage editingGPUImage = new GPUImage(context);
            editingGPUImage.setImage(currentEditingImage);

            if (filterType.equals("aged"))
                bmp = applyAgedEffect(editingGPUImage);
            else if (filterType.equals("blackWhite"))
                bmp = applyBlackWhiteEffect(editingGPUImage);
            else if (filterType.equals("cold"))
                bmp = applyColdEffect(editingGPUImage);
            else if (filterType.equals("rosy"))
                bmp = applyRosyEffect(editingGPUImage);
            else if (filterType.equals("intense"))
                bmp = applyIntenseEffect(editingGPUImage);
            else if (filterType.equals("warm"))
                bmp = applyWarmEffect(editingGPUImage);
            else if (filterType.equals("light"))
                bmp = applyLightEffect(editingGPUImage);

            processPicture(bmp, (float) compressQuality, JPEG, callbackContext);
        }
    }

    private void applyEffectForReview(String pathOrData, final String filterType, final double compressQuality, int isBase64Image, CallbackContext callbackContext) {
        synchronized (this) {
            this.validateInput(pathOrData, filterType, isBase64Image);

            Bitmap bmp = null;
            GPUImage previewGPUImage = new GPUImage(context);
            previewGPUImage.setImage(currentPreviewImage);

            if (filterType.equals("aged"))
                bmp = applyAgedEffect(previewGPUImage);
            else if (filterType.equals("blackWhite"))
                bmp = applyBlackWhiteEffect(previewGPUImage);
            else if (filterType.equals("cold"))
                bmp = applyColdEffect(previewGPUImage);
            else if (filterType.equals("rosy"))
                bmp = applyRosyEffect(previewGPUImage);
            else if (filterType.equals("intense"))
                bmp = applyIntenseEffect(previewGPUImage);
            else if (filterType.equals("warm"))
                bmp = applyWarmEffect(previewGPUImage);
            else if (filterType.equals("light"))
                bmp = applyLightEffect(previewGPUImage);

            processPicture(bmp, (float) compressQuality, JPEG, callbackContext);
        }
    }

    private void applyEffectForThumbnail(final String pathOrData, final String filterType, final double compressQuality, final int isBase64Image, final CallbackContext callbackContext) {
        validateInput(pathOrData, filterType, isBase64Image);

        new Thread(new Runnable() {
            @Override
            public void run() {
                Bitmap thumb = getThumbnailImage();
                Bitmap bmp = null;
                GPUImage thumbnailGPUImage = new GPUImage(context);
                thumbnailGPUImage.setImage(thumb);

                if (filterType.equals("aged"))
                    bmp = applyAgedEffect(thumbnailGPUImage);
                else if (filterType.equals("blackWhite"))
                    bmp = applyBlackWhiteEffect(thumbnailGPUImage);
                else if (filterType.equals("cold"))
                    bmp = applyColdEffect(thumbnailGPUImage);
                else if (filterType.equals("rosy"))
                    bmp = applyRosyEffect(thumbnailGPUImage);
                else if (filterType.equals("intense"))
                    bmp = applyIntenseEffect(thumbnailGPUImage);
                else if (filterType.equals("warm"))
                    bmp = applyWarmEffect(thumbnailGPUImage);
                else if (filterType.equals("light"))
                    bmp = applyLightEffect(thumbnailGPUImage);

                processPicture(bmp, (float) compressQuality, JPEG, callbackContext);
            }
        }).start();
    }

    private Bitmap applyAgedEffect(GPUImage img) {
        return this.applyFilter(img, 0.00516f, 0.04124f, 0.8763f, 0.7474f, 0.1804f, 1.0103f,
                NOT_AVAILABLE, 0.7835f, 0.719f, 0.616f);
    }

    private Bitmap applyBlackWhiteEffect(GPUImage img) {
        return this.applyFilter(img, 0.0f, NOT_AVAILABLE, NOT_AVAILABLE, 1.2282f, 0.2062f, 0.268f,
                NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE);
    }

    private Bitmap applyRosyEffect(GPUImage img) {
        return this.applyFilter(img, 0.79897f, -0.164948f, 0.819588f, 0.881443f, 0.474227f, NOT_AVAILABLE,
                NOT_AVAILABLE, NOT_AVAILABLE, 0.822165f, 0.876289f);
    }

    private Bitmap applyIntenseEffect(GPUImage img) {
        return this.applyFilter(img, 2f, NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE, 0.12371f,
                NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE);
    }

    private Bitmap applyWarmEffect(GPUImage img) {
        return this.applyFilter(img, 1.2577f, -0.085f, 0.964f, 0.8763f, 0.4536f,
                NOT_AVAILABLE, NOT_AVAILABLE, 0.83f, 0.8092f, 0.7938f);
    }

    private Bitmap applyLightEffect(GPUImage img) {
        return this.applyFilter(img, 1.4484f, -0.0592f, 0.7629f, 0.7835f, 0.4124f, -0.0825f,
                NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE, NOT_AVAILABLE);
    }

    private Bitmap applyColdEffect(GPUImage img) {
        return this.applyFilter(img, 0.216495f, -0.134021f, 0.85567f, 1.061856f, 0.603093f, NOT_AVAILABLE,
                NOT_AVAILABLE, 0.708763f, 0.832474f, NOT_AVAILABLE);
    }

    private Bitmap applyFilter(GPUImage img, float saturation, float brightness, float contrast, float gamma, float exposure, float sharpen, float hue,
                               float red, float green, float blue) {

        GPUImageFilterGroup group = new GPUImageFilterGroup();

        if (saturation != NOT_AVAILABLE) {
            GPUImageSaturationFilter saturationFilter = new GPUImageSaturationFilter();
            saturationFilter.setSaturation(saturation);
            group.addFilter(saturationFilter);
        }
        if (brightness != NOT_AVAILABLE) {
            GPUImageBrightnessFilter brightnessFilter = new GPUImageBrightnessFilter();
            brightnessFilter.setBrightness(brightness);
            group.addFilter(brightnessFilter);
        }
        if (contrast != NOT_AVAILABLE) {
            GPUImageContrastFilter contrastFilter = new GPUImageContrastFilter();
            contrastFilter.setContrast(contrast);
            group.addFilter(contrastFilter);
        }
        if (gamma != NOT_AVAILABLE) {
            GPUImageGammaFilter gammaFilter = new GPUImageGammaFilter();
            gammaFilter.setGamma(gamma);
            group.addFilter(gammaFilter);
        }
        if (exposure != NOT_AVAILABLE) {
            GPUImageExposureFilter exposureFilter = new GPUImageExposureFilter();
            exposureFilter.setExposure(exposure);
            group.addFilter(exposureFilter);
        }
        if (sharpen != NOT_AVAILABLE) {
            GPUImageSharpenFilter sharpenFilter = new GPUImageSharpenFilter();
            sharpenFilter.setSharpness(sharpen);
            group.addFilter(sharpenFilter);
        }
        if (hue != NOT_AVAILABLE) {
            GPUImageHueFilter hueFilter = new GPUImageHueFilter();
            hueFilter.setHue(hue);
            group.addFilter(hueFilter);
        }
        if (red != NOT_AVAILABLE || green != NOT_AVAILABLE || blue != NOT_AVAILABLE) {
            GPUImageRGBFilter rgbFilter = new GPUImageRGBFilter();
            if (red != NOT_AVAILABLE)
                rgbFilter.setRed(red);
            if (green != NOT_AVAILABLE)
                rgbFilter.setGreen(green);
            if (blue != NOT_AVAILABLE)
                rgbFilter.setBlue(blue);
            group.addFilter(rgbFilter);
        }

        img.setFilter(group);
        return img.getBitmapWithFilterApplied();
    }

    public void failPicture(String err) {
        this.callbackContext.error(err);
    }

    private void processPicture(Bitmap bitmap, float compressQuality, int encodingType, CallbackContext callbackContext) {
        synchronized (this) {
            ByteArrayOutputStream jpeg_data = new ByteArrayOutputStream();
            CompressFormat compressFormat = encodingType == JPEG ?
                    CompressFormat.JPEG :
                    CompressFormat.PNG;

            try {
                if (bitmap.compress(compressFormat, Math.round(0.8f * 100), jpeg_data)) {
                    byte[] code = jpeg_data.toByteArray();
                    byte[] output = Base64.encode(code, Base64.NO_WRAP);
                    String js_out = new String(output);
                    callbackContext.success(js_out);
                    js_out = null;
                    output = null;
                    code = null;
                }
            } catch (Exception e) {
                this.failPicture("Error compressing image.");
            }
            jpeg_data = null;
        }
    }

    private Bitmap base64ToBitmap(String encodedImage) {
        byte[] decodedString = Base64.decode(encodedImage, Base64.DEFAULT);
        return BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
    }
}
